src/app/borrower/borrower.component.spec.ts

```diff
@@ -1,25 +1,273 @@
+    expect(component.borrower).toEqual(mockBorrower);
+  }));
+
+  it("should be a valid login", () => {
+    component.borrowerLogin.patchValue({ cardNumber: 23 });
+    expect(component.borrowerLogin.valid).toBeTruthy();
+  });
+
+  it("should not be a valid login", () => {
+    component.borrowerLogin.patchValue({ cardNumber: "f43" });
+    expect(component.borrowerLogin.valid).toBeFalsy();
+    component.borrowerLogin.patchValue({ cardNumber: 0 });
+    expect(component.borrowerLogin.valid).toBeFalsy();
+  });
+
+  it("should sort branches", () => {
+    component.availableBooks = [
+      {
+        bookId: 1,
+        branchId: 1,
+        noOfCopies: 0,
+        branchName: "Central Park",
+        bookTitle: "The Adventures of Huckleberry Finn",
+      },
+      {
+        bookId: 1,
+        branchId: 2,
+        noOfCopies: 2,
+        branchName: "Pflugerville",
+        bookTitle: "The Adventures of Huckleberry Finn",
+      },
+      {
+        bookId: 2,
+        branchId: 1,
+        noOfCopies: 1,
+        branchName: "Central Park",
+        bookTitle: "Pride and Prejudice",
+      },
+    ];
+    component.sortAllBranches();
+    expect(component.branches).toEqual(["Central Park", "Pflugerville"]);
+  });
+
+  it("should change pagination count", () => {
+    component.availableBooks = [
+      {
+        bookId: 1,
+        branchId: 1,
+        noOfCopies: 0,
+        branchName: "Central Park",
+        bookTitle: "The Adventures of Huckleberry Finn",
+      },
+      {
+        bookId: 1,
+        branchId: 2,
+        noOfCopies: 2,
+        branchName: "Pflugerville",
+        bookTitle: "The Adventures of Huckleberry Finn",
+      },
+      {
+        bookId: 2,
+        branchId: 1,
+        noOfCopies: 1,
+        branchName: "Central Park",
+        bookTitle: "Pride and Prejudice",
+      },
+    ];
+    component.changePaginationCount();
+    expect(component.filterMetadata.count).toEqual(3);
+  });
+
+  it("should load all books", fakeAsync(() => {
+    let mockBookCopies = [
+      { bookId: 1, branchId: 1, noOfCopies: 0 },
+      { bookId: 4, branchId: 2, noOfCopies: 7 },
+      { bookId: 2, branchId: 3, noOfCopies: 1 },
+    ];
+
+    spyOn(service, "readBookCopies").and.returnValue(
+      Promise.resolve(mockBookCopies)
+    );
+    component.loadAllBooks();
+    tick();
+    expect(service).toBeTruthy();
+    expect(component.availableBooks).toEqual(mockBookCopies);
+  }));
+
+  // it('should return a book', fakeAsync(() => {
+  //   let inputLoan = {bookId: 2, branchId: 5, cardNo: 6,
+  //     dateIn: null,
+  //     dateOut: "2020-04-24T17:58:05.000Z",
+  //     dueDate: "05/11/2020"}
+
+  // }));
+
+  it("should process the borrower loans", fakeAsync(() => {
+    let formattedLoans = [
+      {
+        bookId: 2,
+        branchId: 5,
+        cardNo: 6,
+        dateIn: "2020-06-05T14:36:42.000Z",
+        dateOut: "2020-04-24T17:58:05.000Z",
+        dueDate: "05/11/2020",
+      },
+    ];
```

Comment:
Won't gatherLoanData also add the branch name and book title to the loan?

Reply:
These were some of the challenges I had while testing. I am not sure what I am doing wrong because I figured out how to test some Async functions but whenever the Async happens within the function it won't work. I moved on for time's sake. So this test is only covering a portion of that function.

Reply:
Async stuff is tricky. I've spent longer than I would have liked to debugging my misuse of it.

Looking more closely at the code, it appears the problem might be in processLoan. It looks like it'll set loan.branchName and loan.bookTitle to undefined, because the functions setBranchNameOf and setBookTitleOf have no explicit return values. Though clearly those properties appear to be what they should be in the image you posted to Slack. And similarly for processBookCopy.

When the functions themselves are called, they set the properties to the correct values, but then the assignment statements would make the properties undefined. Maybe it just so happens that the view is reading the properties in between those two events?